### ISSUE TITLE
Loss of data in vector of struct with val(typed_memory_view())

### ISSUE THREAD

**Issue by Ni55aN** [emscripten-core/emscripten#5155]
```
 struct vec9 {
  float a[9];
  float operator[](size_t i) const { return a[i]; };
};

  std::vector<val> vec2valVector(const std::vector<vec9> vector) {
    std::vector<val> v;
    v.push_back(val(typed_memory_view<float>(9, vector[0].a)));
    return v;
  }
```

```
std::vector<vec9> v;
v.push_back(vec9{1, 0, 0, 0, 0.7, 0, 0.2, 0.7, 0});
v.push_back(vec9{1, 0, 0, 0, 0.7, 0, 0.2, 0.7, 0});
return vec2valVector(v)

```
JS return 
> [1, 0, 0, 0, 0.699999988079071, 0, 0.20000000298023224, 0.699999988079071, 0]

But
```
 std::vector<vec9> v;
 v.push_back(vec9{0, 1, 0, 0, 0.7, 0.7, 0.7, 0.7, 0});

vec2valVector(v); 
```

JS return

> [7.360090370917952e-39, 7.36009597611181e-39, 7.36009597611181e-39, 3.5032461608120427e-44, 9.640933434554741e-42, 9.640933434554741e-42, 0.699999988079071, 0.699999988079071, 3.363116314379561e-44]
> 

Why returns an invalid array when there is only one element in the vector?

**Comment by AlexPerrot** [comment]
Hi,
I was able to reproduce your problem.
I am not 100% sure, but here's what I think is happening :

All your data is allocated on the stack and you return memory views to that. But after your functions finish, the data can be replaced by something else on the stack.
If you use data allocated on the heap (i.e. a `vector<vec9>*` or a `vector<vec9*>`), then the data is preserved after the function exits and the data read from the js side is not currupted.

If this is the explanation, this seems to be the result of using memory views on stack data, not a bug in embind.

**Comment by Ni55aN** [comment]
@AlexPerrot, You are right, with pointers everything works fine. But if I use pointers, then I'll have to clear the memory, when to do it? How to understand when JS received them?

**Comment by AlexPerrot** [comment]
It depends on what you want to achieve.
For a general use case, I would just copy the data and not use a memory_view, but you might have other constraints.

**Comment by Ni55aN** [comment]
How can I do this on the C++ side to pass the vector of arrays?

**Comment by AlexPerrot** [comment]
If you want, you can bind `vec9` as a `value_array` and remove the use of memory_view.

**Comment by Ni55aN** [comment]
How this can be done for the array?
Trying 
`value_array<vec9>("vec9").element(&vec9::a);`

I get error

> In file included from /emsdk_portable/emscripten/tag-1.37.3/system/includ
> e/emscripten/bind.h:14:
> In file included from /emsdk_portable/emscripten/tag-1.37.3/system/includ
> e/emscripten/val.h:8:
> /emsdk_portable/emscripten/tag-1.37.3/system/include/emscripten/wire.h:33
> 3:24: error: array 'new' cannot have initialization arguments
>                 return new T(v);
>                        ^     ~
> /emsdk_portable/emscripten/tag-1.37.3/system/include/emscripten/bind.h:52
> 6:39: note: in instantiation of member function 'emscripten::internal::Ge
> nericBindingType<float [9]>::toWireType' requested here
>                 return MemberBinding::toWireType(ptr.*field);
>                                       ^
> /emsdk_portable/emscripten/tag-1.37.3/system/include/emscripten/bind.h:69
> 5:28: note: in instantiation of function template specialization 'emscrip
> ten::internal::MemberAccess<vec9, float [9]>::getWire<vec9>' requested he
> re
>                 ::template getWire<ClassType>;
>                            ^
> reproducebug/em.cpp:55:29: note: in instantiation of function template sp
> ecialization 'emscripten::value_array<vec9>::element<vec9, float [9]>' re
> quested here
>   value_array<vec9>("vec9").element(&vec9::a);
>                             ^
> In file included from reproducebug/em.cpp:1:
> /emsdk_portable/emscripten/tag-1.37.3/system/include/emscripten/bind.h:53
> 5:28: error: array type 'float [9]' is not assignable
>                 ptr.*field = MemberBinding::fromWireType(value);
>                 ~~~~~~~~~~ ^
> /emsdk_portable/emscripten/tag-1.37.3/system/include/emscripten/bind.h:69
> 7:28: note: in instantiation of function template specialization 'emscrip
> ten::internal::MemberAccess<vec9, float [9]>::setWire<vec9>' requested he
> re
>                 ::template setWire<ClassType>;
>                            ^
> reproducebug/em.cpp:55:29: note: in instantiation of function template sp
> ecialization 'emscripten::value_array<vec9>::element<vec9, float [9]>' re
> quested here
>   value_array<vec9>("vec9").element(&vec9::a);

**Comment by AlexPerrot** [comment]
OK, you can't use the default `value_array.element` binding with your data structure.
Other are available : https://github.com/kripken/emscripten/blob/master/system/include/emscripten/bind.h#L735

The main problem you have is that your structure is not assignable, so the only binding that will work is to override the getter and setter.

```
value_array<vec9>("vec9")
        .element(optional_override([](const vec9& v){return v[0];}), optional_override([](const vec9& v, float x){}));
```

And you have to add a `.element` for each of the 9 elements. Tedious, but should work.
I did not implement a setter policy, but you can if you want.
